REGRESSION (Safari 16.4, 258767@main): Carcassonne game on boardgamearena.com unplayable (serialization bug affecting background-position)

darinadler opened the pull request:
#### 82228ed935598ec6fc263134947f46b4d46849f1
<pre>
REGRESSION (Safari 16.4, 258767@main): Carcassonne game on boardgamearena.com unplayable (serialization bug affecting background-position)
<a href="https://bugs.webkit.org/show_bug.cgi?id=256814">https://bugs.webkit.org/show_bug.cgi?id=256814</a>
rdar://109380755

Reviewed by Tim Nguyen.

The code to serialize background-position and mask-position did not correctly handle the case
where the Y value was 0%. Rearranged the code slightly to resolve this.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Added test cases.
* LayoutTests/fast/css/background-position-serialize.html: Ditto.
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt: Ditto.
* LayoutTests/fast/masking/parsing-webkit-mask.html: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-valid-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-valid.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-valid.html:
Updated expectations since initial value of background-position is &quot;0% 0%&quot;, not &quot;0% center&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt: Added test cases.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
Updated expectations since initial value of mask-position is &quot;0% 0%&quot;, not &quot;0% center&quot;.

* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeLayered const): Rearrange the code to correctly serialize
background-position-y and mask-position-y as part of a shorthand.

Canonical link: <a href="https://commits.webkit.org/265056@main">https://commits.webkit.org/265056@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b518ab97d477052fe61a9b5d9ddfe86d462f811f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10135 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11297 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11871 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9847 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12338 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11456 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7933 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8758 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16166 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9027 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8907 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12243 "7 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9397 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7636 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8586 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8662 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2317 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12807 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/css/background-position-serialize.html:
```diff
@@ -99,6 +99,41 @@
 shouldBe("t.style.cssText", "background-position: 20% 20%;");
 shouldBe("t.getAttribute('style')", "background-position: 20% 20%;");
 
+run("t.style.backgroundPositionY = '0%';");
```
Testcases should be added in this WPT as well: css/css-backgrounds/parsing/background-position-valid.html

nt1m commented on LayoutTests/fast/css/background-position-serialize.html:
```diff
@@ -99,6 +99,41 @@
+shouldBe("t.style.backgroundPositionX", "0%");
+shouldBe("t.style.backgroundPositionY", "0%");
+shouldBe("t.style.cssText", "background-position: 0%;");
+shouldBe("t.getAttribute('style')", "background-position: 0%;");
```
This looks incorrect:

https://drafts.csswg.org/css-backgrounds/#background-position

> If only one value is specified, the second value is assumed to be [center](https://drafts.csswg.org/css-backgrounds/#valdef-background-position-center).

So this case should serialize as `background-position: 0% 0%;`.

darinadler replied:
Thanks. I knew there was a reason I added that test case!

darinadler replied:
This second flavor of the bug is caused by a different code path, the "if all are skipped serialize the first" rule.

nt1m reviewed (approved):
Looks good, thanks! Please make sure to export changes to WPT afterwards

darinadler committed:
REGRESSION (Safari 16.4, 258767@main): Carcassonne game on boardgamearena.com unplayable (serialization bug affecting background-position)
https://bugs.webkit.org/show_bug.cgi?id=256814
rdar://109380755

Reviewed by Tim Nguyen.

The code to serialize background-position and mask-position did not correctly handle the case
where the Y value was 0%. Rearranged the code slightly to resolve this.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Added test cases.
* LayoutTests/fast/css/background-position-serialize.html: Ditto.
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt: Ditto.
* LayoutTests/fast/masking/parsing-webkit-mask.html: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-valid-expected.txt: Ditto.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-valid.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-valid.html:
Updated expectations since initial value of background-position is "0% 0%", not "0% center".

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt: Added test cases.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
Updated expectations since initial value of mask-position is "0% 0%", not "0% center".

* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeLayered const): Rearrange the code to correctly serialize
background-position-y and mask-position-y as part of a shorthand.

Canonical link: https://commits.webkit.org/265056@main
